### DISCUSSION TITLE
Compound Chords in 2.2 Alpha

### DISCUSSION THREAD

**Discussion by psoukie** [psoukie/zipchord#155] [Announcements]
<img align="right" src="/psoukie/zipchord/wiki/images/zipchord-compound-chords.gif" width="408" alt="dynamic compound chords animation" />

I'm excited to share that the [redesign](https://github.com/psoukie/zipchord/discussions/145) progressed far enough that I was able to create and release a working proof of concept with "compound chords." (The idea came from the ["context chording" discussion](https://github.com/psoukie/zipchord/discussions/101).)

It allows doing things ZipChord should be able to do: output different text based on the context of previous or following chords (or single key presses). You can see this in the demo. The best thing is, the change will not break how dictionaries and shorthands were defined until now.

Right now, the compound chords are available in an Alpha release. It supports chords, shorthands and hints, but it **does not** support smart spaces and smart capitalization yet. (And there might be some other features that are not yet hooked up to the new architecture.)

The redesign will also allow me to improve other features like the detection and handling of intentional / unintentional chords and shorthands.

If you'd like to experiment, I am attaching the [**zipchord-exe-2.2.0-alpha.zip**](https://github.com/psoukie/zipchord/files/10681656/zipchord-exe-2.2.0-alpha.zip) here, and including two examples how you can define compound chords in your chord dictionaries:

```
tk ;g	taking
da 1	Monday
```
The chords or individual chords and keypresses are separated by spaces.
(It is also still possible to use spaces in chords. The spaces that are part of the chord just need to be at the start of the chord, such as:
```
 n  c	no comment	That's: "[Space]n[Space][Space]c[Tab]no comment"
```
This entry would mean chord "Space+N" followed by chord "Space+C".

_On a personal note: Most of the work on this release was done while my son has been in ICU, during hours when I could not do anything useful and needed to distract my mind with things that require a lot of attention. As of this writing, he's slowly improving, and I hope he will recover quicker than I can complete redesigning the code._

**Comment by nightscape** [comment]
Hi @psoukie, my thought and prayers go out to your son!!!

**Comment by jsampson45** [comment]
Compound chords can have more than two elements. I hope development of this software can continue. It is pretty nifty already.

**Comment by jsampson45** [comment]
However, the same chord cannot be used in different chord sequences. One might have oe pn to give "open" but then one cannot have, as another dictionary entry, pn for "pen". Not that one would want to, but I hope this makes my point clear.

**Reply by psoukie** [reply]
Just a quick note that I finally managed to spend some time back with ZipChord. 

I didn't get to a new releasable version yet, but it shouldn't be too many days now.

Thank you for your patience on this!

**Comment by jsampson45** [comment]
I must apologise because my diagnosis above was wrong. The issue seems to be that "04" is seen as the same as "[space]4". I eliminated all the chords "0" plus a digit from my dictionary file and Zipchord worked OK. Is there a type error?

**Comment by jsampson45** [comment]
I don't know if Zipchord is still being maintained but the facility for adding spaces at the beginnings of chords seems to work only for the first chord for a word.

**Reply by psoukie** [reply]
Thank you for reporting the issues. Other things took over for a few months, but now I went back and fixed some issues in the stable, 2.1 branch and released 2.1.2, and I will resume work on the 2.2 branch, so it can be on par in terms of smart spaces etc. and become a real release.

**Reply by jsampson45** [reply]
Spaces at the beginnings of chords seems to work now so there must have been a glitch.

**Comment by psoukie** [comment]
This was an unintentionally long break, but happy to share that the feature is now officially live in [ZipChord 2.2](https://github.com/psoukie/zipchord/releases/tag/2.2.0)!
